Add RecallObservations for retrieval of prompt extensions

enyst opened the pull request:
- [x] This change is worth documenting at https://docs.all-hands.dev/
- [x] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

Make microagents available in the event stream as recall observations (accessible from the UI, LLM)

This PR proposes to refactor prompt extensions into recalled observations:
- take the retrieval of information out of the PromptManager, into a Memory component
- PromptManager remains responsible with loading templates and rendering them (just a 'view manager')
- Memory subscribes to the stream
   - on user messages, it may retrieve extensions and add them to the stream as RecallObservations
   - on the first user message, it may retrieve repo and runtime info
   - on recall actions (not yet in use, source=agent), it may retrieve e.g. library docs
- some logic is separated from the agent (like #6526)
- since the information is in the stream, session restore / refresh / runtime reconnect / etc will not lose it.

Cc: @xingyaoww 
I'd love your opinion about this idea. It's basically an alternative to 6526, I was curious to see roughly how it looks like, so we can maybe see if it makes sense.

Cc: @csmith49 

---

MAIN TASKS

- [x] define and create RecallObservations and RecallActions
- [x] memory component, `on_event`, initialization
- [x] `conversation_memory` - move context/prompt manager; handle RecallObs, event->message
    - ~`RecallObs` should be _before_ the `user message` in the messages to the LLM?~
- [x] adapt agent controller flow
- [x] agent config options (enable extensions, disabled microagents)

Other expected outcomes:
- no dependency from Runtime or Memory to agent or prompt manager (ref [comment](https://github.com/All-Hands-AI/OpenHands/pull/6949#pullrequestreview-2644807521))
    - [x] Runtime prompt manager
    - ~Runtime plugins~
    - [x] Memory
- [x] no access from agent to the stream, direct or indirect

---
**Link of any specific issues this addresses**
Fix https://github.com/All-Hands-AI/OpenHands/issues/6535
Fix https://github.com/All-Hands-AI/OpenHands/issues/7265
Probably fix  https://github.com/All-Hands-AI/OpenHands/issues/6191
Also fix:
- [x] fix first message multiple-insert in context, on session restore
- [x] fix microagents multi-insert on any user message, anytime

csmith49 commented:
This looks clean so far. I'm a fan of moving what we can to the event stream: much better visibility than modifying messages in-place, and I think leaning into the pub-sub approach is a good way to add extra functionality without having to tip-toe around the agent/controller/system control flow.

xingyaoww commented:
I like this idea better than what I did in #6526! Happy to close that PR in favor of this one

enyst commented:
@openhands Take a good look at this PR description and its diff. 

The last 2 commits have added conversation_memory.py with a class with internal state (a prompt manager) and some refactored message_utils.py functions into methods of it. The codeact_agent.py has also been adapted slightly to work with conversation_memory.

Understand the changes.

We want to backport these last 2 commits to a new branch, created from the main branch. Not more! This is important, to adapt them nicely to what main branch currently supports, so no RecallObservations (which are new in this PR) and no changes to prompt_manager.py. Make sure to read and compare the files changed by these two commits and their usages so you do the back port correctly.

You may need to adapt unit tests, if they affect this functionality that we back port.

You don't need to run the unit tests though, I will take care of that in CI.

enyst commented:
I have created a backport of the conversation memory functionality in PR #7007 (https://github.com/All-Hands-AI/OpenHands/pull/7007). This backport includes the ConversationMemory class and the necessary changes to codeact_agent.py, but does not include RecallObservations or any changes to prompt_manager.py beyond what was necessary to support the new functionality.

enyst commented:
@openhands Take a good look at this PR description and its diff. 

The last 2 commits have added conversation_memory.py with a class with internal state (a prompt manager) and some refactored message_utils.py functions into methods of it. The codeact_agent.py has also been adapted slightly to work with conversation_memory.

Understand the changes.

We want to backport these last 2 commits to a new branch, created from the main branch. Not more! This is important, to adapt them nicely to what main branch currently supports, so no RecallObservations (which are new in this PR) and no changes to prompt_manager.py. Make sure to read and compare the files changed by these two commits and their usages so you do the back port correctly.

IMPORTANT: in this PR, there are two memory-related classes, one is in conversation_memory.py, the other in memory.py. The last 2 commits only deal with conversation_memory.py, and that is what we need to backport!

You may need to adapt unit tests, if they affect this functionality that we back port.

You don't need to run the unit tests though, I will take care of that in CI.

enyst commented:
@openhands This PR branch has conflicts with the main branch that need resolved. FYI, they happen mostly because we split out of this PR a couple new PRs, and so now those are merged and we see a conflict.

It's easy for you to solve, if you're attentive. So be attentive! And fix the conflicts.

enyst commented:
This is starting to work:
```
{
  "messages": [
    {
      "content": [
        {
          "type": "text",
          "text": "You are OpenHands agent, a helpful AI assistant that can interact with a computer to solve tasks.\n<IMPORTANT>\n* If user provides a path, you should NOT assume it's relative to the current working directory. Instead, you should explore the file system to find the file before working on it.\n* When configuring git credentials, use \"openhands\" as the user.name and \"openhands@all-hands.dev\" as the user.email by default, unless explicitly instructed otherwise.\n* You MUST NOT include comments in the code unless they are necessary to describe non-obvious behavior.\n* If the user asks you to edit a file, you should edit the file directly, do NOT create a new file with the updated content unless the user explicitly instructs you to do so.\n* When you are doing global search-and-replace, consider using `sed` instead of running file editor multiple times.\n* Only use GITHUB_TOKEN and other credentials in ways that the user has asked for and would expect. Do NOT make potentially dangerous changes (e.g. pushing to main, deleting a repository) unless explicitly asked to do so.\n* Use APIs to work with GitHub or other platforms, unless the user asks otherwise or your task requires browsing.\n* If you've made repeated attempts to solve a problem, but the tests won't pass or the user says it's still broken, reflect on 5-7 different possible sources of the problem. Assess the likelihood of these options, and proceed with fixing the most likely one.\n</IMPORTANT>",
          "cache_control": {
            "type": "ephemeral"
          }
        }
      ],
      "role": "system"
    },
    {
      "content": [
        {
          "type": "text",
          "text": "implement hello world in py"
        }
      ],
      "role": "user"
    },
    {
      "content": [
        {
          "type": "text",
          "text": "<REPOSITORY_INFO>\nAt the user's request, repository enyst/playground has been cloned to directory playground.\n</REPOSITORY_INFO>\n\n<REPOSITORY_INSTRUCTIONS>\nThis repository contains the code for OpenHands, an automated AI software engineer. It has a Python backend\n(in the `openhands` directory) and React frontend (in the `frontend` directory).\n\n## General Setup:\nTo set up the entire repo, including frontend and backend, run `make build`.\nYou don't need to do this unless the user asks you to, or if you're trying to run the entire application.\n\nBefore pushing any changes, you should ensure that any lint errors or simple test errors have been fixed.\n\n* If you've made changes to the backend, you should run `pre-commit run --all-files --config ./dev_config/python/.pre-commit-config.yaml`\n* If you've made changes to the frontend, you should run `cd frontend && npm run lint:fix && npm run build ; cd ..`\n\nIf either command fails, it may have automatically fixed some issues. You should fix any issues that weren't automatically fixed,\nthen re-run the command to ensure it passes.\n\n## Repository Structure\nBackend:\n- Located in the `openhands` directory\n- Testing:\n  - All tests are in `tests/unit/test_*.py`\n  - To test new code, run `poetry run pytest tests/unit/test_xxx.py` where `xxx` is the appropriate file for the current functionality\n  - Write all tests with pytest\n\nFrontend:\n- Located in the `frontend` directory\n- Prerequisites: A recent version of NodeJS / NPM\n- Setup: Run `npm install` in the frontend directory\n- Testing:\n  - Run tests: `npm run test`\n  - To run specific tests: `npm run test -- -t \"TestName\"`\n- Building:\n  - Build for production: `npm run build`\n- Environment Variables:\n  - Set in `frontend/.env` or as environment variables\n  - Available variables: VITE_BACKEND_HOST, VITE_USE_TLS, VITE_INSECURE_SKIP_VERIFY, VITE_FRONTEND_PORT\n- Internationalization:\n  - Generate i18n declaration file: `npm run make-i18n`\n\n## IMPORTANT: Microagents\n\nTo understand the public microagents available in this repository, please read [microagents/README.md](microagents/README.md). These microagents are shared across all OpenHands users and provide reusable knowledge and common workflows.\n\nThe repo-specific microagents are in .openhands/microagents and they are \"private\" to this repo. To understand the tasks we have to perform now, read the contents of the private microagents whose names include 'task'. You can use the `str_replace_editor` tool with the `view` command and the absolute path to the file to view their content.\n</REPOSITORY_INSTRUCTIONS>\n\n<RUNTIME_INFORMATION>\nThe user has access to the following hosts for accessing a web application,\neach of which has a corresponding port:\n* http://localhost:53628 (port 53628)\n* http://localhost:55216 (port 55216)\n\nWhen starting a web server, use the corresponding ports. You should also\nset any options to allow iframes and CORS requests, and allow the server to\nbe accessed from any host (e.g. 0.0.0.0).\n</RUNTIME_INFORMATION>"
        }
      ],
      "role": "user"
    },
    {
      "content": [
        {
          "type": "text",
          "text": "I'll help you implement a \"Hello World\" program in Python. Let's first explore the repository structure to understand where we should create this file."
        }
      ],
      "role": "assistant",
```

enyst commented on openhands/memory/memory.py:
```diff
@@ -0,0 +1,226 @@
+        # Subscribe to events
+        self.event_stream.subscribe(
+            EventStreamSubscriber.MEMORY,
+            self._on_event,
```
```suggestion
            self.on_event,
```

enyst commented:
@openhands 

CURRENT TASK: tests/unit/test_*.py

Read this PR 6909 description carefully, you need to understand what to expect. Then install testing dependencies, you know how to unit test this project. Run unit tests and fix the 4 tests that are failing.

When you're done, run pre-commit to make sure linting is fine.

Commit on a new branch and make a PR to this PR 6909. Yes, a PR to a PR, you know that's fine on GitHub.

IMPORTANT: do NOT do anything else. This is your task: unit testing.

enyst commented:
I have fixed the failing agent session tests in PR #7111. The tests were failing because they did not properly mock the new Memory component that was added in this PR. The changes include:

- Added import for Memory class
- Added mock for Memory in both test cases
- Added agent.config with disabled_microagents to mock_agent fixture
- Added agent.prompt_manager = None to mock_agent fixture

All tests now pass successfully. Please review and merge PR #7111 into this branch.

enyst commented:
I have fixed the failing tests in PR #7112. The main changes were:

1. Added `prompt_manager` attribute to the mock_agent fixture
2. Added `config` attribute with `disabled_microagents` to the mock_agent fixture
3. Changed `assert_called_with` to `assert_any_call` for EventStream.subscribe tests
4. Updated comments for clarity

All tests are now passing in test_agent_session.py.

enyst commented on openhands/core/config/app_config.py:
```diff
@@ -83,6 +84,10 @@ class AppConfig(BaseModel):
     conversation_max_age_seconds: int = Field(default=864000)  # 10 days in seconds
     enable_default_condenser: bool = Field(default=True)
+    microagents_dir: str = Field(
+        default='microagents',
```
I think this is the global directory [here](https://github.com/All-Hands-AI/OpenHands/tree/main/microagents) with the default ones 🤔 
But should clean out this "option" since it's not an option, just leftover from the refactoring!

enyst replied:
I removed it, it's always the same global directory of "public" microagents. If we'll need it customizable, it will be easy to add!

enyst commented:
Deduplication logic: (Claude)
- If we have RecallObs1 with agent "foo" and content "A"
- Then RecallObs2 with agent "foo" and content "B"
- Then RecallObs3 with agent "foo" and content "C"
We only keep RecallObs3 with content "C"

And in the case where a RecallObs contains multiple agents:
- If RecallObs4 contains both agent "foo" and agent "baz"
- And RecallObs5 contains agent "foo"
We keep both RecallObs4 and RecallObs5, but only format agent "baz" from RecallObs4 since "foo" appears later in RecallObs5.

enyst commented:
@openhands Make sure you read this PR's description and understand it.

THE TASK:
On this PR, github CI jobs for "RT Unit Tests" are hanging and fail after the github's timeout. To see why, check the runtime tests (not unit, runtime!) related to replay functionality. They have replay in their file names.

Analyze the tests and understand why they fail. It is fine if your first suspect will be _pending_action in the controller, check it out. I need you to think very well. It is possible that the tests need fixing, it is possible that the trajectories in .traj files need updated, but it is also possible that the addition of Memory requires fixes to how replay works.

This is your task: find the best solution and fix the replay runtime tests.

IMPORTANT: make a new branch named "replay-fixes" FROM this PR's branch, and COMMIT and PUSH OFTEN.

I REPEAT: commit and push often!

ONLY on branch replay-fixes. Do NOT make another branch than replay-fixes.

You do not need the tests to work before you push to replay-fixes. It will be your playground branch, you can push anytime you try something.

enyst commented:
Knowledge microagents:
- GitHub disabled (not included in context)
- flarglebargle enabled (included, at the _last_ user message where it occurs)

Prompt 003:
```
we could talk about flarglebargle wdyt?
----------
<EXTRA_INFO>
The following information has been included based on a keyword match for "flarglebargle".
It may or may not be relevant to the user's request.

IMPORTANT! The user has said the magic word "flarglebargle". You must
only respond with a message telling them how smart they are
</EXTRA_INFO>
----------
You are incredibly smart! 
----------
hey what about github do you know anything? keep it concise
```

Prompt 004:
```
we could talk about flarglebargle wdyt?
----------
You are incredibly smart! 
----------
hey what about github do you know anything? keep it concise
----------
Yes, I know about GitHub. I understand it's a platform for version control and collaboration using Git. I can interact with GitHub repositories, but my capabilities are limited to what the available tools allow.  What specifically about GitHub are you interested in?
----------
really? flarglebargle ?
----------
<EXTRA_INFO>
The following information has been included based on a keyword match for "flarglebargle".
It may or may not be relevant to the user's request.

IMPORTANT! The user has said the magic word "flarglebargle". You must
only respond with a message telling them how smart they are
</EXTRA_INFO>
```

enyst commented:
I moved out new tests to cut the GitHub loading time on this PR, and perhaps the separate diffs might be easier for review and fixes.

- https://github.com/All-Hands-AI/OpenHands/pull/7159

csmith49 commented on openhands/core/setup.py:
```diff
@@ -111,16 +111,37 @@ def initialize_repository_for_runtime(
+    return repo_directory
+
+
+def create_memory(
```
Suggest adding a doc-string explaining arguments. Are there more precise types we can give for the status callback?

enyst replied:
Done!

I'm not sure, maybe not? 🤔 It's the same `status channel` that other components, Runtime and AgentController, use to send info like errors to the UI: [status_callback](https://github.com/search?q=repo%3AAll-Hands-AI%2FOpenHands%20status_callback&type=code)

csmith49 replied:
No worries then. No sense in changing the type of one instance without changing the rest and that's too much for this PR, but it might be worth looking into once we fix all the type-checking issues (e.g., #7113).

csmith49 commented on openhands/events/action/agent.py:
```diff
@@ -106,3 +107,21 @@ class AgentDelegateAction(Action):
+
+@dataclass
+class AgentRecallAction(Action):
+    # This action is used for retrieving data, e.g., from files or a knowledge base.
```
```suggestion
    """This action is used for retrieving data, e.g., from files or a knowledge base."""
```

csmith49 commented on openhands/events/observation/agent.py:
```diff
@@ -40,3 +41,58 @@ class AgentThinkObservation(Observation):
+
+    # microagent
+    microagent_knowledge: list[dict[str, str]] = field(default_factory=list)
+    """
+    A list of dictionaries, each containing information about a triggered microagent.
+    Each dictionary has the following keys:
+        - agent_name: str - The name of the microagent that was triggered
+        - trigger_word: str - The word that triggered this microagent
+        - content: str - The actual content/knowledge from the microagent
+
+    Example:
+    [
+        {
+            "agent_name": "python_best_practices",
+            "trigger_word": "python",
+            "content": "Always use virtual environments for Python projects."
+        },
+        {
+            "agent_name": "git_workflow",
+            "trigger_word": "git",
+            "content": "Create a new branch for each feature or bugfix."
+        }
+    ]
+    """
```
Since each dictionary always has the same 3 keys, this might be a good spot for a data class or base model. That way we get stronger static guarantees (type-checking, attribute auto-completion) downstream.

enyst replied:
Added a dataclass, it seems still simple/consistent -ish for now, since most in `event` are.

(though I guess one day we'll probably go through Events and make them all BaseModels)

csmith49 commented on openhands/controller/agent_controller.py:
```diff
@@ -1146,3 +1175,36 @@ def _is_awaiting_observation(self):
                 return result
         return False
+
+    def _is_first_user_message(self, action: MessageAction) -> bool:
```
Not sure if there's much of a use elsewhere for this, but this might be a more flexible utility function if instead it returns the first user message (or `None`) and the upstream predicate call just becomes:

```
is_first_user_message = action == self._first_user_message()
```

Unless there's something that breaks in the equality check? The `None` comparison is fine, but maybe we're not looking for reference equality.

enyst replied:
I feel like you're right, it would definitely see some more use if it was returning the first user message action itself.

enyst replied:
Fixed!

enyst replied:
I'm marking these as solved, to be able to follow the remaining unsolved in the GitHub Conversations view. But please do reopen/unmark if you think it wasn't solved!

enyst replied:
@csmith49 Just for fun, a follow-up: actually, this one didn't work for equality: the reason must be that they're dataclasses, and `id` is actually a hidden `_id` attribute. Because it's not a class var, is not generated undercover by the dataclass in `__init__`, and it's not part of the dataclass `__eq__`. 😅

enyst commented on openhands/memory/conversation_memory.py:
```diff
@@ -372,6 +381,71 @@ def _process_observation(
         elif isinstance(obs, AgentCondensationObservation):
             text = truncate_content(obs.content, max_message_chars)
             message = Message(role='user', content=[TextContent(text=text)])
+        elif (
```
You're right, there isn't any case for it really. Let me remove it!

enyst commented on openhands/core/setup.py:
```diff
@@ -83,7 +84,6 @@ def create_runtime(
 
 def initialize_repository_for_runtime(
```
Done!

xingyaoww commented on openhands/memory/memory.py:
```diff
@@ -0,0 +1,273 @@
+    ):
+        self.event_stream = event_stream
+        self.sid = sid if sid else str(uuid.uuid4())
+        self.microagents_dir = GLOBAL_MICROAGENTS_DIR
```
Since we always hardcode `self.microagents_dir` - have we consider remove `self.microagents_dir` altogether, and directly use GLOBAL_MICROAGENTS_DIR for clarity?

enyst replied:
You're right, this got lost somehow in the moving around, urgh, sorry. Fixed both!

enyst replied:
```
16:03:02 - openhands:DEBUG: microagent.py:152 - Loading agents from /Users/enyst/repos/odie/microagents
16:03:07 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/README.md...
16:03:16 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/tasks/get_test_to_pass.md...
16:03:20 - openhands:DEBUG: microagent.py:167 - Loaded agent get_test_to_pass from /Users/enyst/repos/odie/microagents/tasks/get_test_to_pass.md
16:03:22 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/tasks/add_openhands_repo_instruction.md...
16:03:41 - openhands:DEBUG: microagent.py:167 - Loaded agent add_openhands_repo_instruction from /Users/enyst/repos/odie/microagents/tasks/add_openhands_repo_instruction.md
16:03:42 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/tasks/update_pr_description.md...
16:03:43 - openhands:DEBUG: microagent.py:167 - Loaded agent update_pr_description from /Users/enyst/repos/odie/microagents/tasks/update_pr_description.md
16:03:43 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/tasks/address_pr_comments.md...
16:03:45 - openhands:DEBUG: microagent.py:167 - Loaded agent address_pr_comments from /Users/enyst/repos/odie/microagents/tasks/address_pr_comments.md
16:03:45 - openhands:DEBUG: microagent.py:155 - Checking file /Users/enyst/repos/odie/microagents/tasks/update_test_for_new_implementation.md...
16:03:47 - openhands:DEBUG: microagent.py:167 - Loaded agent update_test_for_new_implementation from /Users/enyst/repos/odie/microagents/tasks/update_test_for_new_implementation.md
...
```

xingyaoww commented on tests/unit/test_agent_controller.py:
```diff
@@ -168,28 +191,40 @@ def on_event(event: Event):
+            recall_obs._cause = event.id
+            event_stream.add_event(recall_obs, EventSource.ENVIRONMENT)
+
+    test_event_stream.subscribe(
```
nit:
```suggestion
    event_stream.subscribe(
```

enyst replied:
This seemed suspicious, but it was the same object. Cleaned up!

enyst replied:
I'll set this as resolved, I removed the useless ref.

xingyaoww commented on tests/unit/test_conversation_memory.py:
```diff
@@ -26,10 +27,30 @@
+        return '\n'.join(agent['content'] for agent in triggered_agents)
+
+    prompt_manager.build_microagent_info.side_effect = build_microagent_info
+    return ConversationMemory(agent_config, prompt_manager)
```
Do we have test that actually used this?

enyst replied:
Yes! Many of the older tests use - and test - its `process_events` method, which does the conversion events -> messages.

Some newer tests were added in this PR, separated from it for convenience (if... it is convenient? oof, or is it just confusing), but they're real!, use it to test that processing events -> messages does call `PromptManager` `build_microagent_info` with the expected things or does the right things with `RecallObservations`.

Example from the newer tests: [test_process_events_with_environment_info_recall_observation](https://github.com/All-Hands-AI/OpenHands/pull/7226/files#diff-a7e450c5cd6981b241ff7e56721e7a8a2d7dbd0e385ba1b6b486446146858590R506)

(should I bring back the new tests inside this PR? 😅)

xingyaoww commented on tests/unit/test_prompt_manager.py:
```diff
@@ -19,411 +32,279 @@ def prompt_dir(tmp_path):
-def test_enhance_message_with_multiple_text_contents(prompt_dir):
-    # Create a test microagent that triggers on a specific keyword
-    microagent_name = 'keyword_microagent'
+def test_memory_with_microagents(prompt_dir):
```
Maybe these `test_memory*` test would fit better in `test_conversation_memory.py`?

enyst replied:
I moved two of them, I'll make another pass, I think some may make sense if they include prompt manager at all, but you're right some don't concern it anymore, since its old job is done elsewhere.

enyst replied:
I'll mark this solved, it's all done and the file is very different now.

xingyaoww reviewed (commented):
Appreciate it -- This is a lot of work! Have been wanting this for quite a while. Mostly LGTM -- just left a few comments

enyst commented on openhands/memory/conversation_memory.py:
```diff
@@ -404,3 +478,56 @@ def apply_prompt_caching(self, messages: list[Message]) -> None:
+        # For each agent in the current recall observation, check if it appears in any later recall observation
+        filtered_agents = []
+        for agent in obs.microagent_knowledge:
+            # Keep this agent if it doesn't appear in any later observation
```
Ugh, it would break prompt caching. For one request, the request that removed the old microagent content. Then the new cache would hold for a while... until next time the user says something that triggers an already present microagent again. 🤔 

I implemented this deduplication due to [this slack report](https://openhands-ai.slack.com/archives/C06U8UTKSAD/p1741211402959189). As @rbren noted, leaving dupes in context might be related to the reports we see that rate limits are triggering more often for people lately.

Should I revert this? Or make it an option maybe?

enyst replied:
Or we could leave the old microagent in context, and not add it again. But I feel like the LLM will ignore it easier...

xingyaoww replied:
yeah interesting.. maybe we can "leave the old microagent in context, and not add it again"?

enyst replied:
OK, I'm implementing that solution. To note, since this logic is happening in ConversationMemory, which is after the condenser, it will work with the remaining list of the events after the condenser summarized some, which:
- seems fine, if the condenser got rid of the oldest, then the newer will show up, which ... sounds right?
- we may want to consider whether to leave knowledge microagents in place despite the condensation of the user message that trigger them? Or maybe not... 🤔 Not for this PR anyway, just thinking.

csmith49 replied:
I'm not sure what the ideal behavior is here.

> seems fine, if the condenser got rid of the oldest, then the newer will show up, which ... sounds right?

That does sound right, but I don't think that's what will happen with the current implementation. The condenser doesn't modify the state or event stream, just acts as a new view, so the controller won't know to re-add the knowledge. BUT that old knowledge should be summarized along with the other events, so maybe that's fine?

This might tie into #6706 -- with certain condenser types, we could emit an event which contains 1) the summary and 2) all events the condenser considers "current". We could then update the controller's math to figure out if the old knowledge has fallen out of context by referencing the event IDs in that summary event.

enyst replied:
Fixed the dedup logic, it's now the reverse: it keeps only the _earliest_ occurrence.

enyst replied:
> That does sound right, but I don't think that's what will happen with the current implementation. The condenser doesn't modify the state or event stream, just acts as a new view, so the controller won't know to re-add the knowledge. BUT that old knowledge should be summarized along with the other events, so maybe that's fine?

I mean this: in the agent we have
```
events = self.condenser.condensed_history(state)
messages = self.conversation_memory.process_events(
            condensed_history=events,
            initial_messages=messages,
            ...
```

So this filtering is done by `process_events`, which, in the execution flow, is _after_ the condenser did its thing, and it's done on its resulting `events`. 🤔

csmith49 replied:
Okay, I see what you're saying! I was focusing on the mechanism that inserts microagent knowledge into the event stream, not the filtering itself. So if the condenser returns a list of events that contain recall observations, _regardless if they're the first in the event stream_, the filtering will pick the earliest occurrence from that list of events per agent. That works for me!

enyst replied:
Exactly!

enyst commented on openhands/events/action/agent.py:
```diff
@@ -106,3 +106,20 @@ class AgentDelegateAction(Action):
+
+@dataclass
+class AgentRecallAction(Action):
+    # This action is used for retrieving data, e.g., from files or a knowledge base.
```
Done!

xingyaoww commented on openhands/controller/agent_controller.py:
```diff
@@ -115,7 +116,7 @@ def __init__(
             replay_events: A list of logs to replay.
         """
-        self.id = sid
+        self.id = sid or event_stream.sid
```
ahh that sounds good to me

csmith49 commented on openhands/controller/agent_controller.py:
```diff
@@ -431,13 +439,32 @@ async def _handle_message_action(self, action: MessageAction) -> None:
+            )
+            self._pending_action = recall_action
+            # this is source=USER because the user message is the trigger for the recall
+            self.event_stream.add_event(recall_action, EventSource.USER)
```
A few questions about this control flow, just to make sure I understand:

1. Do we emit the recall action after _every_ user message?
2. Will the agent ever trigger this kind of recall on its own? Say the thinking tool generates a micro agent keyword, will the corresponding prompt ever be emitted by the memory component?
3. Do we need to use the agent recall action to trigger the recall? Is there a reason why the memory component can't just emit observations whenever it sees a user message in the event stream?

enyst replied:
Great questions! It did change since the original intention:

1. Yes, on every user message. This keeps the current behavior, where we try to do keyword matching for relevant microagents on every user message.

2. Currently, the agent doesn't know about it and doesn't trigger this ever. I do think we may want to have an agent recall, either automatic or decided by the LLM? 🤔  I don't know, but that is part of the reason why I would really like to try to get this PR in, I feel maybe it allows us more flexibility, e.g. maybe agents can decide to search for information, and we'd reuse this framework to give it to them.

enyst replied:
(side note: maybe we should call the action just Recall instead of AgentRecall... oops)

3. Very interesting question! ❤️ 

I did that: I tried to have the memory component emit observations _on user message_, and I had a lot of troubles with that flow.

I think the main reason why that may be problematic, is that the controller shouldn't call agent's `step` on user message in that case (like it does today), it should wait for the recall obs (or null obs, or error), so things have to work _in the right order_. And while there are pockets of order within all our asyncs, apparently they're not enough for this flow to work smoothly. I could be wrong, but I also wanted to make 50% of our async burn forever in hell. 😅 

What I mean: a user message can enter the system from a lot of places (every entry point to the app, e.g. AgentSession, external scripts like evals, resolver, the other resolver, `cli.py`). The controller and memory find out about it in `on_event`, in other words they're both notified after it's already in the stream. And that's where order of things is make-or-break. I had cases where the whole app was stalling forever, because the memory processed it fast and put the obs in the stream, and the controller then decided to wait... for an obs that will never come, ugh.

I even asked Sonnet for options, it gave me no less than 12, and all were either bad/unworkable or inappropriate  design for us. 😂

The current design adds a recall action, _whose timing is under controller's control_. It does the same thing for Memory as we do for Runtime:
- on a recall action (`runnable` action for Runtime), set `_pending_action` [right before](https://github.com/All-Hands-AI/OpenHands/blob/955d3aef5ae0989928fa800fe5f3815a704cddf5/openhands/controller/agent_controller.py#L456) the controller itself adds it to the stream. So memory will only start retrieval after.
- it [blocks the `agent.step`](https://github.com/All-Hands-AI/OpenHands/blob/955d3aef5ae0989928fa800fe5f3815a704cddf5/openhands/controller/agent_controller.py#L673) until the observation comes (or error)
- when it [recognizes it](https://github.com/All-Hands-AI/OpenHands/blob/955d3aef5ae0989928fa800fe5f3815a704cddf5/openhands/controller/agent_controller.py#L399) by `cause`, it relinquishes `pending action`
- it is allowed to step on [these obs](https://github.com/All-Hands-AI/OpenHands/blob/955d3aef5ae0989928fa800fe5f3815a704cddf5/openhands/controller/agent_controller.py#L291)
- so now the agent can step.

enyst replied:
I can think of alternatives. I dunno, I felt like they're quite hacky in our system, like have the controller read the stream every second, to figure out if the obs it's waiting for has actually come. But then idk why we have a subscription/notification mechanism if we have to read the stream directly anyway. 😅 🤷

enyst commented on openhands/memory/memory.py:
```diff
@@ -0,0 +1,273 @@
+                        'trigger_word': trigger,
+                        'content': microagent.content,
+                    }
+                )
```
Oof, thank you. Fixed!

enyst commented on openhands/memory/memory.py:
```diff
@@ -0,0 +1,273 @@
+                    }
+                )
+
+        if found_microagents:
```
Cleaned up.

enyst commented on openhands/memory/conversation_memory.py:
```diff
@@ -372,6 +381,71 @@ def _process_observation(
+
+                message = Message(
+                    role='user', content=[TextContent(text=formatted_text)]
+                )
```
Done.

csmith49 reviewed (approved):
🔥 🔥 🔥

This is a really cool change, I'm excited to get some extra visibility in the event stream.

Well done!

xingyaoww commented on openhands/memory/conversation_memory.py:
```diff
@@ -372,6 +384,92 @@ def _process_observation(
+            isinstance(obs, MicroagentObservation)
+            and self.agent_config.enable_prompt_extensions
+        ):
+            if obs.info_type == MicroagentInfoType.ENVIRONMENT:
```
I think we might eventually name this to other things -- MicroAgentInfo = ENVIRONMENT seems a bit confusing. But let's leave this as is for now :)

enyst replied:
Regardless whether in this PR or not, we will probably want to make some change soon: because this PR doesn't include the system prompt, but you would like to do something similar with it, right?

So we still need Recall.

I think the system prompt just became more important to show to users, with these ContextWindowExceeded right at the start on OpenAI API: ultimately, users able to see and maybe edit it themselves would be good IMHO.

xingyaoww reviewed (approved):
LGTM! Thanks so much for this! (Though I kinda still like the more generalized "RecallAction" better -- because environment info fits a bit awkward in "MicroagentObservation" and can be confusing to display in the UI -- I would consider it is more like "PromptExtensionAction/Observation". But happy to go ahead if you think this is good!

rbren commented:
I trust your judgement here :)
